Have PlatformCALayerRemoteCustom plumb preservesFlip to UIProcess

nmoucht opened the pull request:
#### 3912f312cb1e86980a0630241041c5ddc80c189f
<pre>
Have PlatformCALayerRemoteCustom plumb preservesFlip to UIProcess
<a href="https://bugs.webkit.org/show_bug.cgi?id=247341">https://bugs.webkit.org/show_bug.cgi?id=247341</a>
rdar://101786702

Reviewed by Tim Horton and Simon Fraser.

Since the effects of setGeometryFlipped don&apos;t cross a hosting boundary,
it is not propogated to the PDFKit layers being hosted in the WebProcess
when UI-side compositing is on. This causes rendering to look wrong and
scrolling in the wrong direction. To solve this issue, we plumb a bit through
to the UIProcess to set preservesFlip, which allows the value of geometryFlipped
to propogate through layer hosting. We also add a check for ui-side compositing
when creating PDF scrollbars due to a bug where the scrollbars didn&apos;t reflect the
geometryFlipped value of its ancestor.

* Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h:
* Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.mm:
(+[CALayer _web_renderLayerWithContextID:shouldPreserveFlip:]):
(+[CALayer _web_renderLayerWithContextID:]): Deleted.
* Source/WebKit/Platform/cocoa/LayerHostingContext.mm:
(WebKit::LayerHostingContext::createPlatformLayerForHostingContext):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::LayerCreationProperties::LayerCreationProperties):
(WebKit::RemoteLayerTreeTransaction::LayerCreationProperties::encode const):
(WebKit::RemoteLayerTreeTransaction::LayerCreationProperties::decode):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::enterAcceleratedCompositingMode):
(WebKit::PageClientImpl::didFirstLayerFlush):
(WebKit::PageClientImpl::updateAcceleratedCompositingMode):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteCustom.mm:
(WebKit::PlatformCALayerRemoteCustom::populateCreationProperties):

Canonical link: <a href="https://commits.webkit.org/257379@main">https://commits.webkit.org/257379@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/773713ad291e6c6f90ecec0e3454292d3b16bfb4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98525 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7735 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31656 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107949 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168220 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102473 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8246 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85123 "Build is in progress. Recent messages:Pull request contains relevant changes; Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91069 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/105952 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104195 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6259 "Found 2 new test failures: css3/supports-dom-api.html, fast/text/text-edge-no-half-leading-simple.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89817 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33248 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88077 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21176 "Found 2 new test failures: css3/supports-dom-api.html, fast/text/text-edge-no-half-leading-simple.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76188 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1671 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22705 "Found 2 new test failures: css3/supports-dom-api.html, fast/text/text-edge-no-half-leading-simple.html (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1591 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45204 "Found 4 new test failures: css3/supports-dom-api.html, fast/text/text-edge-with-margin-padding-border-simple.html, ipc/pasteboard-write-custom-data.html, media/video-inaccurate-duration-ended.html (failure)") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/5072 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6519 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42129 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2559 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2969 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
```diff
@@ -48,6 +48,7 @@
     , type(WebCore::PlatformCALayer::LayerTypeLayer)
     , hostingContextID(0)
     , hostingDeviceScaleFactor(1)
+    , preservesFlip(0)
```
surely "false"

nmoucht replied:
Fixed.

hortont424 commented on Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.mm:
```diff
@@ -79,13 +79,14 @@ - (void)_web_setLayerTopLeftPosition:(CGPoint)position
     [layerHost setName:@"Hosting layer"];
 #endif
     layerHost.contextId = contextID;
+    layerHost.preservesFlip = preservesFlip;
```
is it harmful to make this always true? (I can't tell if it's "preserve geometryFlipped, which might only be set in the PDFPlugin case" or "this will always cause a flip")

nmoucht replied:
This is only set true for PlatformCALayerRemoteCustom. And setting this value doesn't actually flip the geometry, it just preserves the value that's set across layer hosting boundaries.

smfr commented on Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h:
```diff
@@ -27,7 +27,7 @@
 @interface CALayer (WebCoreCALayerExtras)
 
-+ (CALayer *)_web_renderLayerWithContextID:(uint32_t)contextID;
++ (CALayer *)_web_renderLayerWithContextID:(uint32_t)contextID shouldPreserveFlip:(bool)preservesFlip;
```
Argument should be the Objective-C type `BOOL` not `bool`

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/Platform/cocoa/LayerHostingContext.mm:
```diff
@@ -93,7 +93,7 @@
 RetainPtr<CALayer> LayerHostingContext::createPlatformLayerForHostingContext(LayerHostingContextID contextID)
 {
-    return [CALayer _web_renderLayerWithContextID:contextID];
+    return [CALayer _web_renderLayerWithContextID:contextID shouldPreserveFlip:false];
```
`shouldPreserveFlip:NO`

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
```diff
@@ -118,6 +118,7 @@ class RemoteLayerTreeTransaction {
 
         uint32_t hostingContextID;
         float hostingDeviceScaleFactor;
+        bool preservesFlip;
```
As a followup, it would be nice to use C++ initializers in this class.

smfr commented on Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
```diff
@@ -599,15 +599,15 @@ - (BOOL)_hostsLayersInWindowServer;
     ASSERT(!layerTreeContext.isEmpty());
 
-    CALayer *renderLayer = [CALayer _web_renderLayerWithContextID:layerTreeContext.contextID];
+    CALayer *renderLayer = [CALayer _web_renderLayerWithContextID:layerTreeContext.contextID shouldPreserveFlip:false];
```
`shouldPreserveFlip:false` -> `shouldPreserveFlip:NO` here and below

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteCustom.mm:
```diff
@@ -104,6 +104,7 @@
     PlatformCALayerRemote::populateCreationProperties(properties, context, type);
     properties.hostingContextID = hostingContextID();
     properties.hostingDeviceScaleFactor = context.deviceScaleFactor();
+    properties.preservesFlip = true;
```
= YES

nmoucht replied:
Fixed.

smfr reviewed (approved):
Can we make a layout test to show that this change works?

smfr commented on Source/WebKit/WebProcess/Plugins/PluginView.h:
```diff
@@ -98,6 +98,8 @@ class PluginView final : public WebCore::PluginViewBase {
     WebCore::FloatRect rectForSelectionInRootView(PDFSelection *) const;
     CGFloat contentScaleFactor() const;
+    
+    bool isUsingUISideCompositing();
```
Should be `const`

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
```diff
@@ -1369,11 +1374,12 @@ static void dataProviderReleaseInfoCallback(void* info)
 Ref<Scrollbar> PDFPlugin::createScrollbar(ScrollbarOrientation orientation)
 {
     auto widget = Scrollbar::createNativeScrollbar(*this, orientation, ScrollbarControlSize::Regular);
+    auto shouldFlip = static_cast<PluginView *>(static_cast<RenderEmbeddedObject *>(m_view->pluginElement().renderer())->widget())->isUsingUISideCompositing();
```
Yikes, that's quite the chain of casts. We should use `downcast<>` or `dynamicDowncast<>` here instead.

smfr commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -1579,6 +1579,8 @@ class WebPage : public API::ObjectImpl<API::Object::Type::BundlePage>, public IP
 
     void generateTestReport(String&& message, String&& group);
+    
+    bool isUsingUISideCompositing() { return m_drawingAreaType == DrawingAreaType::RemoteLayerTree; }
```
Doesn't build on all platforms.

nmoucht committed:
Have PlatformCALayerRemoteCustom plumb preservesFlip to UIProcess
https://bugs.webkit.org/show_bug.cgi?id=247341
rdar://101786702

Reviewed by Tim Horton and Simon Fraser.

Since the effects of setGeometryFlipped don't cross a hosting boundary,
it is not propogated to the PDFKit layers being hosted in the WebProcess
when UI-side compositing is on. This causes rendering to look wrong and
scrolling in the wrong direction. To solve this issue, we plumb a bit through
to the UIProcess to set preservesFlip, which allows the value of geometryFlipped
to propogate through layer hosting. We also add a check for ui-side compositing
when creating PDF scrollbars due to a bug where the scrollbars didn't reflect the
geometryFlipped value of its ancestor.

* Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.h:
* Source/WebCore/platform/graphics/cocoa/WebCoreCALayerExtras.mm:
(+[CALayer _web_renderLayerWithContextID:shouldPreserveFlip:]):
(+[CALayer _web_renderLayerWithContextID:]): Deleted.
* Source/WebKit/Platform/cocoa/LayerHostingContext.mm:
(WebKit::LayerHostingContext::createPlatformLayerForHostingContext):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::LayerCreationProperties::LayerCreationProperties):
(WebKit::RemoteLayerTreeTransaction::LayerCreationProperties::encode const):
(WebKit::RemoteLayerTreeTransaction::LayerCreationProperties::decode):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::enterAcceleratedCompositingMode):
(WebKit::PageClientImpl::didFirstLayerFlush):
(WebKit::PageClientImpl::updateAcceleratedCompositingMode):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteCustom.mm:
(WebKit::PlatformCALayerRemoteCustom::populateCreationProperties):

Canonical link: https://commits.webkit.org/257379@main
